pkg/controller/acmechallenges/scheduler/sync.go

```diff
@@ -0,0 +1,113 @@
+)
+
+// Sync will process a single ACME challenge resource in order to determine
+// whether it can be scheduled for processing.
+// This is currently extremelly primitive, and **will not** do intelligent
+// things like bumping challenges that are for already expired or nearing expiry
+// certificates to the 'front' of the queue.
+//
+// This may be something to do in future - we could use a resyncFunc to build
+// a stack of challenges to process, and upon observation of a new challenge,
+// re-evaluate the whole stack.
+//
+// For now, this function will simply be used to solve https://github.com/jetstack/cert-manager/issues/951
+func (c *Controller) Sync(ctx context.Context, ch *cmapi.Challenge) error {
+	// If the challenge is already in a final state, there is nothing more for
+	// us to do.
+	if acme.IsFinalState(ch.Status.State) {
+		return nil
+	}
+
+	// If the challenge already has 'processing' set to true, there is nothing
+	// more for us to do.
+	// The 'acmechallenges' controller is responsible for setting this field to
+	// false once processing has completed.
+	if ch.Status.Processing == true {
+		return nil
+	}
+
+	// Begin the scheduling algorithm! Here, we must evaluate all challenges
+	// currently in the apiserver, and their current state, in order to determine
+	// whether we can begin processing this challenge.
+
+	allChallenges, err := c.challengeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	// First, filter out all challenges that are *not* being processed.
+	// With our naive scheduling algorithm, we only care about avoiding *duplicate*
+	// challenges occurring at once.
+	inFlightChallenges := removeNotProcessingChallenges(allChallenges)
+
+	// if any other challenges are in-flight with the same challenge type and
+	// same dnsName, we will *not* mark this challenge as processing
+	for _, inFCh := range inFlightChallenges {
+		if ch.Spec.DNSName == inFCh.Spec.DNSName && ch.Spec.Type == inFCh.Spec.Type {
+			return fmt.Errorf("another %q challenge for challenge %q (domain %q) is in progress, waiting until it is complete before processing", ch.Spec.Type, ch.Name, ch.Spec.DNSName)
+		}
+	}
+
+	// if there are no 'conflicts' detected above, then we can mark this challenge
+	// as processing.
+	ch.Status.Processing = true
+	_, err = c.CMClient.CertmanagerV1alpha1().Challenges(ch.Namespace).Update(ch)
+	if err != nil {
+		return err
+	}
+
+	// we ignore the return value from waitForCacheSync - if it is false, the
+	// controller will shutdown anyway.
+	_ = c.waitForCacheSync()
+
+	return nil
+}
+
+// removeNotProcessingChallenges will filter out challenges from the given slice
+// that have status.processing set to false.
+// TODO: we currently call this function on every call to Sync().
+// In large deployments, this could cause high CPU and memory consumption as it
+// works at O(n^2) complexity (i.e. for every challenge, we have to touch every
+// challenge).
```

Comment:
I wonder how okay this is going to be at the (totally random) 72k certificate scale

Reply:
👍 yep, my thinking here was we can implement the trivial approach and optimise from there